### ISSUE TITLE
Initialization of [EC]DH Parameters

### ISSUE THREAD

**Issue by firetwister** [ghaderer/libmicrohttpd#2]
When using GnuTLS it was not neccessary to initialize the DH or ECDH Paramters explicitly and specifying the ciphers with MHD_OPTION_*_PRIORITIES was sufficient. 
When using Openssl they have to be initialized, or those ciphers will not work. 
If you for example would choose the [modern](https://wiki.mozilla.org/Security/Server_Side_TLS) cipher suite as recommended by Mozilla no TLS connection is possible. Using MHD_OPTION_HTTPS_MEM_DHPARAMS does work, but there is currently no option to set ECDH params. Once I crudely hacked that initialization TLS was working fine.
```
EC_KEY *ecdh=EC_KEY_new_by_curve_name(NID_X9_62_prime256v1)
SSL_CTX_set_tmp_ecdh(context->d.openssl.context, ecdh);
EC_KEY_free(ecdh);
```

**Comment by ghaderer** [comment]
I guess adding a MHD_OPTION_HTTPS_MEM_ECDHPARAMS option would do the trick. Does it sound as a good solution to you? Should we also try to automatically set a reasonable default value to remove the need to explicitly set this option? What's your opinion on this?

**Comment by ghaderer** [comment]
I see the function `SSL_CTX_set_ecdh_auto()` that removes the need for explicit configuration.

**Comment by ghaderer** [comment]
It's only available since version 1.0.2.

**Comment by firetwister** [comment]
I was thinking of adding the MEM_ECDHPARAMS option, but from what I've seen (not much) it's quite uncommon to have ecdh params stored in files, but call them by name (openssl ecparam -list_curves). I don't know how GnuTLS handles it. On the other hand, all the other MHD_HTTPS options work this way, and not by name.

SSL_CTX_set_ecdh_auto() looks good to me. Openssl version 1.0.2. is the oldest supported
[version](https://www.openssl.org/policies/releasestrat.html), so why support unsupported versions?
This function was removed in openssl 1.1.0, because ECDH is always enabled. However compatibility macros are there.
https://github.com/openssl/openssl/issues/1437
So SSL_CTX_set_ecdh_auto() actually looks perfect, because nobody had the need to set ecdh params with the GnuTLS implementation for years, and it would make both implementations easier interchangeable.

**Comment by ghaderer** [comment]
I agree with you. I think it's important it works out-of-the-box to be able to easily switch between different TLS implementations. I'll give SSL_CTX_set_ecdh_auto() a try.

Thanks for you comments.

**Comment by firetwister** [comment]
I just tested SSL_CTX_set_ecdh_auto() and it works fine.
With Openssl 1.1.0 on the other hand it will not compile ATM, but for other reasons (bio stuff and the ciphers in MHD_TLS_openssl_get_session_cipher_algorithm()). 
[OpenSSL_1.1.0_Changes](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes)

**Comment by ghaderer** [comment]
It could reproduce the issue you mentioned on my side. I made the changes for automatic curve selection. I tested on my side. It works now and I saw no regression with other algorithms.

Regarding the build failure with OpenSSL 1.1, I'll take a look at it.

Thanks again!
